src/components/Router/useWorkflowSteps.tsx

```diff
@@ -0,0 +1,203 @@
+import { StepConfig } from '~types/steps'
+import { useCallback, useEffect, useRef, useState } from 'preact/hooks'
+
+import { formatStep } from '../../index'
+import { NarrowSdkOptions, UrlsConfig } from '~types/commons'
+import {
+  StepsLoadingStatus,
+  StepsHook,
+  CompleteStepValue,
+} from '~types/routers'
+import { poller, PollFunc, Engine } from '../WorkflowEngine'
+import type { WorkflowResponse } from '../WorkflowEngine/utils/WorkflowTypes'
+import useUserConsent from '~contexts/useUserConsent'
+
+type WorkflowStepsState = {
+  status: StepsLoadingStatus
+  steps: StepConfig[]
+  taskId: string | undefined
+  error: string | undefined
+}
+
+const defaultState: WorkflowStepsState = {
+  status: 'idle',
+  taskId: undefined,
+  error: undefined,
+  steps: [],
+}
+
+export const createWorkflowSteps = (
+  { token, workflowRunId, ...options }: NarrowSdkOptions,
+  { onfido_api_url }: UrlsConfig
+): StepsHook => () => {
+  const { addUserConsentStep } = useUserConsent()
+
+  //@todo: We should move the logic of enforcing the `welcome` step from index.ts here
+  const [state, setState] = useState<WorkflowStepsState>({
+    ...defaultState,
+    steps: options.steps,
+  })
+
+  useEffect(() => {
+    // We only inject this step in the first workflow task
+    if (options.steps.every(({ type }) => type !== 'welcome')) {
+      return
+    }
+
+    setState((state) => ({
+      ...state,
+      steps: addUserConsentStep(options.steps),
+    }))
+  }, [addUserConsentStep])
+
+  const { taskId, status, error, steps } = state
+
+  const docData = useRef<Array<unknown>>([])
+  const personalData = useRef({})
+
+  const pollStep = useCallback((cb: () => void) => {
+    if (!token) {
+      throw new Error('No token provided')
+    }
+
+    if (!workflowRunId) {
+      throw new Error('No workflowRunId provided')
```

Comment:
We have exceptions and callbacks mixed. The exception wouldn't be caught is my guess as it's thrown inside a useCallback. Better to use the callback with an error parameter.